File: lab_3_C#/Storage.cs
using System;
using System.Collections.Generic;
using System.Threading;

public class Storage
{
    private Queue<int> buffer = new Queue<int>();
    private int capacity;

    private Semaphore emptySlots;
    private Semaphore fullSlots;
    private Semaphore mutex = new Semaphore(1, 1);

    public Storage(int capacity)
    {
        this.capacity = capacity;
        emptySlots = new Semaphore(capacity, capacity);
        fullSlots = new Semaphore(0, capacity);
    }

    public void Produce(int item, int producerId)
    {
        emptySlots.WaitOne();
        mutex.WaitOne();

        buffer.Enqueue(item);
        Console.WriteLine($"Producer {producerId} produced: {item}");

        mutex.Release();
        fullSlots.Release();
    }

    public int Consume(int consumerId)
    {
        fullSlots.WaitOne();
        mutex.WaitOne();

        int item = buffer.Dequeue();
        Console.WriteLine($"Consumer {consumerId} consumed: {item}");

        mutex.Release();
        emptySlots.Release();

        return item;
    }
}

File: lab_3/src/model/Storage.java
package model;

import java.util.LinkedList;
import java.util.Queue;
import java.util.concurrent.Semaphore;

public class Storage {
    private final Queue<Integer> buffer = new LinkedList<>();
    private final int capacity;

    private final Semaphore emptySlots;
    private final Semaphore fullSlots;
    private final Semaphore mutex;

    public Storage(int capacity) {
        this.capacity = capacity;
        this.emptySlots = new Semaphore(capacity);
        this.fullSlots = new Semaphore(0);
        this.mutex = new Semaphore(1);
    }

    public void produce(int item, int producerId) throws InterruptedException {
        emptySlots.acquire();
        mutex.acquire();

        buffer.add(item);
        System.out.println("Producer " + producerId + " produced: " + item);

        mutex.release();
        fullSlots.release();
    }

    public int consume(int consumerId) throws InterruptedException {
        fullSlots.acquire();
        mutex.acquire();

        int item = buffer.remove();
        System.out.println("Consumer " + consumerId + " consumed: " + item);

        mutex.release();
        emptySlots.release();
        return item;
    }
}
    